react/the_react_ecosystem/react_router.md

```diff
@@ -0,0 +1,589 @@
+
+React Router is a standard routing library for React applications. By using React Router, we can specify React components, that can be rendered based on the route, and so much more. Let's dive in!
+
+### Adding A Router
+
+Let's make a small app to understand how this router is implemented. Create a new project using `create-react-app` and get rid of the boilerplate code as you already learned in previous lessons (just leave the `index.js` and the `App.js` files). Let's add some mock pages as an example. Create a new `Profile.js` file with the following component:
+
+~~~jsx
+const Profile = () => {
+  return (
+    <div>
+      <h1>Hello from profile page!</h1>
+      <p>So, how are you?</p>
+    </div>
+  );
+};
+
+export default Profile;
+~~~
+
+Replace the `App.js` file with some basic content too:
+
+~~~jsx
+const App = () => {
+  return (
+    <div>
+      <h1>Hello from the main page of the app!</h1>
+      <p>Here are some examples of links to other pages</p>
+      <nav>
+        <ul>
+          <li>
+            <a href="profile">Profile page</a>
+          </li>
+        </ul>
+      </nav>
+    </div>
+  );
+};
+
+export default App;
+~~~
+
+Now it's time to add the router! There's a couple of ways of defining our app's routes, but in **React Router v6.7.0 or higher**, it is recommended to add routes as objects. Add the following to `root.js`, we will talk about what is happening in a little bit.
+
+~~~jsx
+import React from "react";
+import ReactDOM from "react-dom/client";
+import { createBrowserRouter, RouterProvider } from "react-router-dom";
+import App from "./App";
+import Profile from "./Profile";
+
+const router = createBrowserRouter([
+  {
+    path: "/",
+    element: <App />,
+  },
+  {
+    path: "/profile",
+    element: <Profile />,
+  },
+]);
+
+ReactDOM.createRoot(document.getElementById("root")).render(
+  <React.StrictMode>
+    <RouterProvider router={router} />
+  </React.StrictMode>
+);
+~~~
+
+Once this is done, go ahead and run `npm start` and check out both routes: the home route `/` and the profile route `/profile` It works! But what is happening here?
+
+1. We import `createBrowserRouter` and `RouterProvider` from React Router.
+2. `createBrowserRouter` is used to create the configuration for a router by simply passing arguments in the form of an array of routes.
+3. The configuration array contains objects with two mandatory keys, the path and the corresponding element to be rendered.
+4. This generated configuration is then rendered in, by passing it to the `RouterProvider` component.
+
+### The Link Element
+
+But you may notice, when we click the links in the navbar, the browser is reloading for the next URL instead of using React Router. This isn't what was promised! To help with this, React Router exports a custom `Link` element to be used instead of the regular `a` tag. We can replace the `a` tag in our navbar with the `Link` element.
```

Comment:
We could a reference link for Link:
```suggestion
But you may notice, when we click the links in the navbar, the browser is reloading for the next URL instead of using React Router. This isn't what was promised! To help with this, [React Router exports a custom `Link` element](https://reactrouter.com/en/main/components/link) to be used instead of the regular `a` tag. We can replace the `a` tag in our navbar with the `Link` element.
```